tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala

```diff
@@ -0,0 +1,179 @@
+package io.tofhir.engine.data.write
+
+import io.tofhir.engine.config.ToFhirConfig
+import io.tofhir.engine.model._
+import io.tofhir.engine.util.FileUtils
+import io.tofhir.engine.util.FileUtils.FileExtensions
+import org.apache.hadoop.fs.FileUtil
+import org.apache.spark.sql.functions.{col, from_json, schema_of_json}
+import org.apache.spark.sql.{Dataset, SaveMode, SparkSession}
+
+import java.io.File
+import java.net.URI
+import java.nio.file.{Files, Path, Paths, StandardCopyOption}
+import java.util
+
+/**
+ * Spark archiver for writing data sources of failed mappings to configured folder
+ */
+object SparkArchiver {
+
+  /**
+   * Archive data sources row by row based on error types (e.g. invalid input, mapping error, invalid resource) to configured folder
+   *
+   * @param spark               spark session to create dataset from not written resources
+   * @param mappingJobExecution job execution to get output directory of data sources with errors
+   * @param mappingUrl          to create directory for each mapping url within the job execution
+   * @param notWrittenResources created fhir resources that cannot be written to fhir server
+   * @param mappingErrors       data sources that got error because of the mapping definition
+   * @param invalidInputs       data sources that got error because of invalid input
+   */
+  def archiveDataSourcesOfFailedMappings(spark: SparkSession,
+                                         mappingJobExecution: FhirMappingJobExecution,
+                                         mappingUrl: Option[String],
+                                         notWrittenResources: util.List[FhirMappingResult],
+                                         mappingErrors: Dataset[FhirMappingResult],
+                                         invalidInputs: Dataset[FhirMappingResult]): Unit = {
+    if (mappingJobExecution.job.dataProcessingSettings.saveErroneousRecords) {
+      if (!invalidInputs.isEmpty) {
+        this.writeDatasetToConfiguredFolder(mappingJobExecution, invalidInputs, mappingUrl.get, FhirMappingErrorCodes.INVALID_INPUT)
+      }
+      if (!mappingErrors.isEmpty) {
+        this.writeDatasetToConfiguredFolder(mappingJobExecution, mappingErrors, mappingUrl.get, FhirMappingErrorCodes.MAPPING_ERROR)
+      }
+      if (!notWrittenResources.isEmpty) {
+        import spark.implicits._
+        val notWrittenResourcesDs = spark.createDataset[FhirMappingResult](notWrittenResources)
+        this.writeDatasetToConfiguredFolder(mappingJobExecution, notWrittenResourcesDs, mappingUrl.get, FhirMappingErrorCodes.INVALID_RESOURCE)
+      }
+    }
+  }
+
+  /**
+   * Writes the dataset to the configured folder
```

Comment:
What is the configured folder here? I suppose there are subfolders w.r.t. the error type. If so, I recommend to describe this better in the javadoc.